lib/Compiler.js

```diff
@@ -2,520 +2,524 @@
+			loader: null,
+			context: null
+		};
+		let deprecationReported = false;
+		this.parser = {
+			plugin: (hook, fn) => {
+				if(!deprecationReported) {
+					console.warn("webpack: Using compiler.parser is deprecated.\n" +
+						"Use compiler.plugin(\"compilation\", function(compilation, data) {\n  data.normalModuleFactory.plugin(\"parser\", function(parser, options) { parser.plugin(/* ... */); });\n}); instead. " +
+						"It was called " + new Error().stack.split("\n")[2].trim() + ".");
+					deprecationReported = true;
+				}
+				this.plugin("compilation", (compilation, data) => {
+					data.normalModuleFactory.plugin("parser", parser => {
+						parser.plugin(hook, fn);
+					});
 				});
-			});
-		}.bind(this),
-		apply: function() {
-			var args = arguments;
-			if(!deprecationReported) {
-				console.warn("webpack: Using compiler.parser is deprecated.\n" +
-					"Use compiler.plugin(\"compilation\", function(compilation, data) {\n  data.normalModuleFactory.plugin(\"parser\", function(parser, options) { parser.apply(/* ... */); });\n}); instead. " +
-					"It was called " + new Error().stack.split("\n")[2].trim() + ".");
-				deprecationReported = true;
-			}
-			this.plugin("compilation", function(compilation, data) {
-				data.normalModuleFactory.plugin("parser", function(parser) {
-					parser.apply.apply(parser, args);
+			},
+			apply: () => {
+				const args = arguments;
+				if(!deprecationReported) {
+					console.warn("webpack: Using compiler.parser is deprecated.\n" +
+						"Use compiler.plugin(\"compilation\", function(compilation, data) {\n  data.normalModuleFactory.plugin(\"parser\", function(parser, options) { parser.apply(/* ... */); });\n}); instead. " +
+						"It was called " + new Error().stack.split("\n")[2].trim() + ".");
+					deprecationReported = true;
+				}
+				this.plugin("compilation", (compilation, data) => {
+					data.normalModuleFactory.plugin("parser", parser => {
+						parser.apply.apply(parser, args);
+					});
 				});
-			});
-		}.bind(this)
-	};
-
-	this.options = {};
-}
-module.exports = Compiler;
+			}
+		};
 
-Compiler.prototype = Object.create(Tapable.prototype);
-Compiler.prototype.constructor = Compiler;
+		this.options = {};
+	}
 
-Compiler.Watching = Watching;
-Compiler.prototype.watch = function(watchOptions, handler) {
-	this.fileTimestamps = {};
-	this.contextTimestamps = {};
-	var watching = new Watching(this, watchOptions, handler);
-	return watching;
-};
+	static Watching(compiler, watchOptions, handler) {
+		return new Watching(compiler, watchOptions, handler);
+	}
 
-Compiler.prototype.run = function(callback) {
-	var self = this;
-	var startTime = Date.now();
+	watch(watchOptions, handler) {
+		this.fileTimestamps = {};
+		this.contextTimestamps = {};
+		const watching = new Watching(this, watchOptions, handler);
+		return watching;
+	}
 
-	self.applyPluginsAsync("before-run", self, function(err) {
-		if(err) return callback(err);
+	run(callback) {
+		const self = this;
```

Comment:
converted 'self' to 'this' in the run method. converted old style function declarations into arrow functions and had to shift position of the function block within the method due to hoisting concerns I think.